### PR TITLE
Spring hw 18 solution

### PR DESCRIPTION
**Pull request by AnatolyNz**

### CODE REVIEW

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.jdbc.datasource.init.ScriptUtils;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc // Automatically configure MockMvc
```
Comment (okuzan) [inline src/test/java/mate/academy/controller/BookControllerTest.java]:
```suggestion
@AutoConfigureMockMvc
```
remove GPT comments :))

File: src/test/java/mate/academy/repository/BookRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    @Sql(scripts = {
+            "classpath:database/books/add-books-to-books-table.sql",
+            "classpath:database/books/add-category-to-categories-table.sql",
+            "classpath:database/books/add-category-to-book.sql"
+    }, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {
+            "classpath:database/books/remove-from-book_category.sql",
+            "classpath:database/books/remove-from-books.sql",
+            "classpath:database/books/remove-from-categories.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
```
Comment (okuzan) [inline src/test/java/mate/academy/repository/BookRepositoryTest.java]: setup is the same for each method, extract to a class level

File: src/test/java/mate/academy/repository/CategoryRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    private static final int PAGE_NUMBER = 0;
+    private static final int PAGE_SIZE = 2;
+
+    @Autowired
+    private CategoryRepository categoryRepository;
+
+    @Test
+    @DisplayName("""
+       Get category by valid ID
+            """)
```
Comment (okuzan) [inline src/test/java/mate/academy/repository/CategoryRepositoryTest.java]:
```suggestion
    @DisplayName("Get category by valid ID")
```
fix everywhere for short names

File: src/test/java/mate/academy/service/CategoryServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    private CategoryRepository categoryRepository;
+
+    @InjectMocks
+    private CategoryServiceImpl categoryService;
+
+    @Test
+    @DisplayName("""
+       Verify search() for all books
+            """)
+    public void testFindAll_ValidSearchParameters_ShouldReturnListOfCategories() {
```
Comment (okuzan) [inline src/test/java/mate/academy/service/CategoryServiceTest.java]:
```suggestion
    public void findAll_ValidSearchParameters_ShouldReturnListOfCategories() {
```

File: src/test/java/mate/academy/service/CategoryServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        assertEquals(expectedDto, actualDto);
+        assertEquals(updatedCategoryDto.getName(), existingCategory.getName());
+        assertEquals(updatedCategoryDto.getDescription(), existingCategory.getDescription());
+    }
+
+    @Test
+    @DisplayName("""
+       Verify update() updated categories with valid ID
+            """)
+    public void testUpdateById_WithNotValidCategoryId_ShouldThrowException() {
```
Comment (okuzan) [inline src/test/java/mate/academy/service/CategoryServiceTest.java]:
```suggestion
    public void updateById_WithNotValidCategoryId_ShouldThrowException() {
```
unify naming

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job, clean code :) minor remarks

File: src/main/java/mate/academy/service/BookServiceImpl.java [hunk 6]
```diff
@@ -42,10 +42,12 @@ public void deleteById(Long id) {
     }
 
     @Override
-    public void updateById(Long id, CreateBookRequestDto bookWithoutId) {
+    public BookDto updateById(Long id, CreateBookRequestDto bookWithoutId) {
         Book book = bookMapper.toModel(bookWithoutId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/service/BookServiceImpl.java]: It's better to 
```suggestion
        Book book = bookRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException(getNotFoundMessage(id)));
        bookMapper.updateBookFromDto(bookDto, book);
```
where in BookMapper
`void updateBookFromDto(CreateBookRequestDto book, @MappingTarget Book entity);`

Comment (Elena-Bruyako) [reply src/main/java/mate/academy/service/BookServiceImpl.java]: Should we set categories here? 
```
book.setCategories(categoriesIdToCategories(bookDto.categoriesId()));
return bookMapper.toDto(bookRepository.save(book));
```

File: src/test/java/mate/academy/repository/BookRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        expected.setAuthor("Author 1");
+        expected.setIsbn("ISBN-123456");
+        expected.setDescription("Description for Book 1");
+        expected.setCoverImage("image1.jpg");
+        expected.setDeleted(false);
+        Optional<Book> actualOptional = bookRepository.getBookById(1L);
+
+        Assertions.assertTrue(actualOptional.isPresent(), "Book with id 1 should be present");
+        Book actual = actualOptional.get();
+        Assertions.assertEquals(expected.getId(), actual.getId());
```
Comment (Elena-Bruyako) [inline src/test/java/mate/academy/repository/BookRepositoryTest.java]: Let's use import static
```suggestion
        assertEquals(expected.getId(), actual.getId());
```
Check all places

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! let's consider some comments.

File: src/main/java/mate/academy/dto/CreateBookRequestDto.java [hunk 8]
```diff
@@ -2,19 +2,24 @@
...
 public class CreateBookRequestDto {
     @NotBlank
     private String title;
     @NotBlank
     private String author;
     @NotBlank
     private String isbn;
-    @NotBlank
+    @NotNull
     @Min(0)
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/dto/CreateBookRequestDto.java]:
```suggestion
    @Positive
```

Comment (MentorMateAcademy) [reply src/main/java/mate/academy/dto/CreateBookRequestDto.java]: Better use this anno

File: src/main/java/mate/academy/repository/book/BookRepository.java [hunk 9]
```diff
@@ -21,4 +21,3 @@ public interface BookRepository extends JpaRepository<Book, Long>,
 
     List<Book> findAll();
 }
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/repository/book/BookRepository.java:19]: Method save already exists in JpaRepository, check others
image

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+                .andReturn();
+
+        //Then
+        BookDto actual = objectMapper.readValue(result.getResponse()
+                .getContentAsString(), BookDto.class);
+        assertNotNull(actual);
+        assertNotNull(actual.getId());
+        assertEquals(expected.getTitle(), actual.getTitle());
+        assertEquals(expected.getPrice(), actual.getPrice());
+    }
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/controller/BookControllerTest.java]: image

Comment (MentorMateAcademy) [reply src/test/java/mate/academy/controller/BookControllerTest.java]: when and then use in correct places

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    @DisplayName("Find all books")
+    void findAll_GivenBookInCatalog_ShouldReturnAllBooks() throws Exception {
+        //Given
+        List<BookDto> expected = new ArrayList<>();
+        expected.add(new BookDto().setId(1L).setTitle("Book 1").setAuthor("Author 1")
+                .setIsbn("ISBN-123456").setPrice(BigDecimal.valueOf(100)).setCategoryIds(Set.of()));
+        expected.add(new BookDto().setId(2L).setTitle("Book 2").setAuthor("Author 2")
+                .setIsbn("ISBN-654321").setPrice(BigDecimal.valueOf(200)).setCategoryIds(Set.of()));
+        expected.add(new BookDto().setId(3L).setTitle("Book 3").setAuthor("Author 3")
+                .setIsbn("ISBN-908765").setPrice(BigDecimal.valueOf(250)).setCategoryIds(Set.of()));
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/controller/BookControllerTest.java]: write better ib stack and Strings what you use write to constants and reuse it

File: src/test/java/mate/academy/controller/CategoryControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+                .setDescription("Description for Book 2")
+                .setCoverImage("image2.jpg"));
+        expected.add(new BookDtoWithoutCategoryIds()
+                .setId(3L)
+                .setTitle("Book 3")
+                .setAuthor("Author 3")
+                .setIsbn("ISBN-908765")
+                .setPrice(BigDecimal.valueOf(250))
+                .setDescription("Description for Book 3")
+                .setCoverImage("image3.jpg"));
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/controller/CategoryControllerTest.java]: Use list.of and better create this templates in private methods or util class

File: src/test/java/mate/academy/repository/BookRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        assertTrue(actualOptional.isPresent(), "Book with id 1 should be present");
+        Book actual = actualOptional.get();
+        assertEquals(expected.getId(), actual.getId());
+        assertEquals(expected.getTitle(), actual.getTitle());
+        assertEquals(expected.getPrice(), actual.getPrice());
+        assertEquals(expected.getAuthor(), actual.getAuthor());
+        assertEquals(expected.getIsbn(), actual.getIsbn());
+        assertEquals(expected.getDescription(), actual.getDescription());
+        assertEquals(expected.getCoverImage(), actual.getCoverImage());
+        assertEquals(expected.isDeleted(), actual.isDeleted());
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/repository/BookRepositoryTest.java]: You don`t need to compare every field if you can compare this objects

File: src/test/java/mate/academy/service/BookServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+@ExtendWith(MockitoExtension.class)
+public class BookServiceTest {
+    private static final String BOOK_AUTHOR_1 = "Author 1";
+    private static final String BOOK_TITLE_1 = "Book 1";
+    private static final String BOOK_PRICE_1 = "300";
+    @Mock
+    private BookRepository bookRepository;
+
+    @InjectMocks
+    private BookServiceImpl bookService;
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/service/BookServiceTest.java]: Place InjectMocks after Mock

File: src/test/java/mate/academy/service/BookServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+        when(bookMapper.toModel(bookRequestDto)).thenReturn(book);
+        when(bookMapper.toDto(savedBook)).thenReturn(expectedDto);
+
+        // Mock repository behavior
+        when(bookRepository.save(book)).thenReturn(savedBook);
+
+        // Test
+        BookDto actualDto = bookService.save(bookRequestDto);
+
+        // Assertions
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/service/BookServiceTest.java]: Don`t need this comments for code

File: src/test/java/mate/academy/service/BookServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    @Test
+    @DisplayName("Verify the correct book was returned when book exists")
+    public void getBookById_WithValidBookId_ShouldReturnValidBook() {
+        Long id = 1L;
+        Book book = new Book();
+        book.setId(id);
+        when(bookRepository.getBookById(id)).thenReturn(Optional.of(book));
+
+        BookDto expected = new BookDto();
+        expected.setId(id);
```
Comment (MentorMateAcademy) [inline src/test/java/mate/academy/service/BookServiceTest.java:80]: use this code in start of method

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/repository/book/BookRepository.java [hunk 17]
```diff
@@ -12,12 +12,8 @@
 public interface BookRepository extends JpaRepository<Book, Long>,
         JpaSpecificationExecutor<Book> {
 
-    Book save(Book book);
-
     @Query("SELECT b FROM Book b JOIN b.categories c WHERE c.id=:categoryId")
     List<Book> findAllByCategoryId(Long categoryId);
 
     Optional<Book> getBookById(Long id);
```
Comment (sarakhmen) [inline src/main/java/mate/academy/repository/book/BookRepository.java:18]: do you really need your custom method for this operation? Please use OOTB methods of JpaRepository instead

**Review by sarakhmen — APPROVED** [review]
Good job!

**Pull request merged**